### DISCUSSION TITLE
`jsonable_encoder` returns objects that cannot be encoded as JSON

### DISCUSSION THREAD

**Discussion by Jonas1312** [fastapi/fastapi#11462] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import json
from dataclasses import dataclass
from enum import Enum

from fastapi.encoders import jsonable_encoder


@dataclass(frozen=True, kw_only=True)
class SomeDataclass:
    a: str
    b: int


class SomeEnum(Enum):
    foo = SomeDataclass(a="foo", b=1)
    bar = SomeDataclass(a="bar", b=2)


temp = jsonable_encoder(SomeEnum.foo)

json.dumps(temp)
```

### Description

Hi,

When executing the code above, I encounter a `TypeError: Object of type SomeDataclass is not JSON serializable` error.

Perhaps the implementation should pass `obj.value` to `jsonable_encoder`:  https://github.com/tiangolo/fastapi/blob/3cc5efc5dee2bf31364d5820581f972197f94a83/fastapi/encoders.py#L257

The parameter `custom_encoder` solves my problem, but I'm wondering if this behavior is expected, as I assume that anything returned by `jsonable_encoder` should be JSON-encodable.

Thanks :)

### Operating System

Windows

### FastAPI Version

0.110.1

### Pydantic Version

2.6.3

### Python Version

Python 3.12.3

**Comment by sehraramiz** [comment]
for now it's looks like (according to jsonable_encoder implementation) you should pass ```SomeEnum.foo.value```
but i think you are right maybe  [this part](https://github.com/tiangolo/fastapi/blob/3cc5efc5dee2bf31364d5820581f972197f94a83/fastapi/encoders.py#L257) of jsonable_encoder should change to something like this:

```python
    if isinstance(obj, Enum):
        return jsonable_encoder(
            obj.value,
            include=include,
            exclude=exclude,
            by_alias=by_alias,
            exclude_unset=exclude_unset,
            exclude_defaults=exclude_defaults,
            exclude_none=exclude_none,
            custom_encoder=custom_encoder,
            sqlalchemy_safe=sqlalchemy_safe,
        )
```

**Reply by Jonas1312** [reply]
Yes passing the enum value directly would work indeed.

But this solution is not ideal if your enum is inside another object:

```python
temp = jsonable_encoder({"some_key": SomeEnum.foo})
```

An enum can also be a dictionary key:

```python
{SomeEnum.foo: "some_value"}
```

So yes I wonder if the current implementation should be changed, since an Enum value can be anything (though it's often a string or an int, but not always).
